File: 0400_Nth_Digit/solution.java
package cn.huang.leetcode;

/*
400. Nth Digit
Find the nth digit of the infinite integer sequence 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, ...

Note:
n is positive and will fit within the range of a 32-bit signed integer (n < 231).

Example 1:

Input:
3

Output:
3
Example 2:

Input:
11

Output:
0

Explanation:
The 11th digit of the sequence 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, ... is a 0, which is part of the number 10.
 */
public class LC_0400_NthDigit {
    /*
    这里有这样一个规律，从1到9，有9个只占1个位数的数字，数字个数为10 - 1 + 1；从10到99，有90个占2个位数的数字，数字个数为99 - 10 + 1
    1——9	   1	9-1+1 = 9	9*1=9
    10——99	   2	99-10+1=90	90*2=180
    100——999   3	999-100+1=900	900*3=2700

    https://www.youtube.com/watch?v=3aGHrMBfO1o
     */
    public int findNthDigit(int n) {
        // 该范围内所有实际数字都占用了digit个单数字
        int digit = 1;
        // 该范围之前的所有实际数字已经占用了totalDigit个单数字
        long totalDigit = 0;
        // 先查出区间范围
        while (true) {
            long top = totalDigit + digit * 9 * (long) Math.pow(10, digit - 1);
            if (n >= totalDigit && n <= top)
                break;
            totalDigit = top;
            digit++;
        }

        // 根据范围算出具体的值
        int start = (int) Math.pow(10, digit - 1);
        int ret = 0;
        totalDigit += 1;
        // 第n个digit在哪个具体的实际数字上
        int value = start + (n - (int) totalDigit) / digit;
        ret = Integer.toString((int) value).charAt((int) ((n - totalDigit) % digit)) - '0';
        return ret;
    }

    // https://www.youtube.com/watch?v=IvJuM8Dr9E8
    public int findNthDigit_2(int n)
    {
        // len   数字的长度, 1位数，2位数...
        // start 同样长度的数字的最小值 1, 10, 100
        // total 同样长度的数字有多少，9， 90，
        for(long len =1, start =1, total=9;;)
        {
            if(n <=len * total)
            {
                long num = (n -1)/ len + start;
                int ret = Integer.toString((int)num).charAt((n - 1) % (int)len) - '0';
                return ret;
            }
            else
            {
                n -= len * total;
                start *= 10;
                len ++;
                total = total * 10;
            }
        }
    }
}

File: 0400_Nth_Digit/solution.cpp
/*
400. Nth Digit

Level: Medium

https://leetcode.com/problems/nth-digit
*/
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

class Solution
{
public:
  int findNthDigit(int n)
  {
    // 开始时是一位数
    long long range_number_length = 1;
    // 开始时一位数有 9 个
    long long range = 9;     
    long long num = 1;

    while (n > range * range_number_length)
    {
      n -= range * range_number_length;
      ++range_number_length;

      num += range;   
      range *= 10;
    }
    //当n落到某一个确定的区间里了，那么(n-1)/len就是目标数字在该区间里的坐标，加上num就是得到了目标数字，
    // 然后我们将目标数字start转为字符串，(n-1)%len就是所要求的目标
    num += (n - 1) / range_number_length;
    return to_string(num).at((n - 1) % range_number_length) - '0';
  }
};